### PR TITLE
EPPT-2588: Fire Severity Index: Add a BuildUpIndex class to IMPROVER

### PR DESCRIPTION
**Pull request by mo-philrelton**
[EPPT-2588](https://metoffice.atlassian.net/browse/EPPT-2588)

In order to calculate the daily Fire Severity Index, we require a Buildup Index calculation. This class, and associated tests, partially reproduce the Canadian Forest Fire Weather Index from [van Wagner and Pickett's 1985 FORTRAN implementation](https://metoffice.sharepoint.com/sites/ProgrammesEnhancingPostProcessingProject/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2FProgrammesEnhancingPostProcessingProject%2FShared%20Documents%2Fscience%5Fworkflows%2FFSI%5FVanWagner%5Fand%5FPickett%5F1985%2Epdf&parent=%2Fsites%2FProgrammesEnhancingPostProcessingProject%2FShared%20Documents%2Fscience%5Fworkflows)

Testing:

- [x] Ran tests and they passed OK
- [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by Anzerkhan27 — APPROVED** [review]
The implementation is consistent with the existing fire weather plugins, the structure aligns with the `FireWeatherIndexBase`. The tests provide good coverage

File: improver/fire_weather/build_up_index.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        From Van Wagner and Pickett (1985), Page 7: Equations 27a-27b.
+
+        Returns:
+            np.ndarray: The calculated BUI values.
+        """
+        dmc_data = self.input_dmc.data
+        dc_data = self.input_dc.data
+
+        # Condition 1: If both DMC and DC are zero, set BUI = 0
+        both_zero = (dmc_data == 0.0) & (dc_data == 0.0)
```
Comment (MoseleyS) [inline improver/fire_weather/build_up_index.py]: Floating point equalities are flaky. It would be better to use `np.isclose(dmc_data, 0.0, atol=1e-7)`, or use the if statement in the Fortran: `dmc_data < 1e-7`

File: improver/fire_weather/build_up_index.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        # Condition 2: If DMC <= 0.4 * DC use equation 27a
+        use_27a = dmc_data <= 0.4 * dc_data
```
Comment (MoseleyS) [inline improver/fire_weather/build_up_index.py:59]: The Fortran has `<` instead of `<=`. I doubt this will have any noticeable impact.

File: improver/fire_weather/build_up_index.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            ) * (0.92 + (0.0114 * dmc_data) ** 1.7)
+
+        # Apply conditions using np.where:
+        # 1. If both_zero: BUI = 0
+        # 2. Elif use_27a (DMC <= 0.4*DC): BUI = bui_27a
+        # 3. Else (DMC > 0.4*DC): BUI = bui_27b
+        bui = np.where(both_zero, 0.0, np.where(use_27a, bui_27a, bui_27b))
+
+        # Ensure BUI is never negative
+        bui = np.clip(bui, 0.0, None)
```
Comment (MoseleyS) [inline improver/fire_weather/build_up_index.py:78]: I think that if condition 1 is changed to match the Fortran inequality, it will mean that bui can never be negative and this line becomes superfluous.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    Verifies BUI calculation from DMC and DC values.
+
+    Args:
+        dmc_val (float): DMC value to test.
+        dc_val (float): DC value to test.
+        expected_bui (float): Expected BUI value.
+    """
+    cubes = input_cubes(dmc_val=dmc_val, dc_val=dc_val)
+    plugin = BuildUpIndex()
+    plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    dmc_values = np.array([0.0, 5.0, 10.0, 20.0, 50.0, 100.0])
+    dc_values = np.array([0.0, 10.0, 30.0, 100.0, 300.0, 500.0])
+
+    for dmc in dmc_values:
+        for dc in dc_values:
+            cubes = input_cubes(dmc_val=dmc, dc_val=dc)
+            plugin = BuildUpIndex()
+            plugin.load_input_cubes(CubeList(cubes))
+            bui = plugin._calculate()
+            assert np.all(bui >= 0.0), f"Negative BUI for DMC={dmc}, DC={dc}"
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: This test does not check that the `np.clip` statement correctly zeroes any negative values, because none of the input pairs result in negative values from the equations. I also cannot come up with any values that can produce negative values because negative inputs are rejected by the plugin and complex inputs are simplified by the submethods of `input_cubes`.

Therefore I think this test should be removed.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+    # Verify unique values (no broadcast errors)
+    assert len(np.unique(bui)) > 1
+
+    # Check specific positions using the equations
+    # Position [0,0]: DMC=5, DC=10, use eq 27b
+    assert np.allclose(bui[0, 0], 4.90, rtol=0.02)
+
+    # Position [2,2]: DMC=60, DC=120, use eq 27b
+    assert np.allclose(bui[2, 2], 59.84, rtol=0.02)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py:128]: There is no spatial processing in the plugin, so why do we include this test?

File: improver_tests/fire_weather/test_build_up_index.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+    # Verify increasing DMC and DC both increase BUI (generally)
+    # Compare bottom-right vs top-left
+    assert result.data[2, 2] > result.data[0, 0]
+
+    # Verify unique values (no broadcast errors)
+    assert len(np.unique(result.data)) > 1
+
+    # Check that different environmental conditions produce different outputs
+    assert not np.allclose(result.data[0, 0], result.data[2, 2], atol=1.0)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py:198]: Again - I don't think this test is necessary.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+    cubes = [
+        make_cube(dmc_values, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_values, "drought_code", "1"),
+    ]
+
+    result = BuildUpIndex().process(CubeList(cubes))
+
+    # When DC=0, BUI = DMC - [0.92 + (0.0114*DMC)^1.7]
+    assert np.allclose(result.data, expected_bui, rtol=1e-3)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: This should be an extra case in `test__calculate`:
```
       # Case 9; DC=0, BUI should be close to DMC
        (60.0, 0.0, 58.56),
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+    cubes = [
+        make_cube(dmc_values, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_values, "drought_code", "1"),
+    ]
+
+    result = BuildUpIndex().process(CubeList(cubes))
+
+    # When DMC=0, BUI should be 0 (no fuel moisture contribution)
+    assert np.allclose(result.data, 0.0, atol=1e-6)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: This should also be in `test__calculate`, and indeed it already is.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+def test_process_both_zero() -> None:
+    """Test that when both DMC and DC are zero, BUI equals 0."""
+    cubes = input_cubes(dmc_val=0.0, dc_val=0.0)
+    result = BuildUpIndex().process(CubeList(cubes))
+
+    assert isinstance(result, Cube)
+    assert result.long_name == "build_up_index"
+    assert result.units == "1"
+    assert np.allclose(result.data, 0.0, atol=1e-6)
+    assert result.dtype == np.float32
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: This is also more appropriate in `test__calculate`, and is already there.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    """
+    cubes = input_cubes(dmc_val=10.0, dc_val=30.0)
+    plugin = BuildUpIndex()
+    plugin.load_input_cubes(CubeList(cubes))
+
+    # Check that attributes exist and have correct values
+    assert hasattr(plugin, "input_dmc")
+    assert hasattr(plugin, "input_dc")
+    assert np.allclose(plugin.input_dmc.data, 10.0)
+    assert np.allclose(plugin.input_dc.data, 30.0)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py:216]: This is pretty much already covered by `test_fire_weather_index_base.py`, but I don't mind much if it is left here too.

File: improver_tests/fire_weather/test_build_up_index.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+    Args:
+        dmc_val (float): DMC value for all grid points.
+        dc_val (float): DC value for all grid points.
+        expected_error (str): Expected error message substring.
+    """
+    cubes = input_cubes(dmc_val, dc_val)
+    plugin = BuildUpIndex()
+
+    with pytest.raises(ValueError, match=expected_error):
+        plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: Move this to be an extra case in `test_validate_input_raises_error`?

File: improver_tests/fire_weather/test_build_up_index.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+        dmc_val = np.nan
+    elif invalid_input_type == "input_dmc_inf":
+        dmc_val = np.inf
+    elif invalid_input_type == "input_dc_nan":
+        dc_val = np.nan
+    elif invalid_input_type == "input_dc_inf":
+        dc_val = -np.inf
+
+    cubes = input_cubes(dmc_val, dc_val)
+    plugin = BuildUpIndex()
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py:285]: Again, as a case in the base class test module

File: improver_tests/fire_weather/test_build_up_index.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+    # Process should complete without warnings since output stays in valid range
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")  # Turn warnings into errors
+        result = plugin.process(CubeList(cubes))
+
+    # No warnings should have been raised (if any were, they'd be errors)
+    assert isinstance(result, Cube)
+    # Verify output is within expected range (0-500 for BUI)
+    assert np.all(result.data >= 0.0)
+    assert np.all(result.data <= 500.0)
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py:313]: This is covered by the base-class tests, although I note that there isn't coverage for when a warning IS generated. Please can you add one?

File: improver_tests/fire_weather/test_build_up_index.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+    """Test that BUI calculation never produces negative values."""
+    # Test a range of DMC and DC values
+    dmc_values = np.array([0.0, 5.0, 10.0, 20.0, 50.0, 100.0])
+    dc_values = np.array([0.0, 10.0, 30.0, 100.0, 300.0, 500.0])
+
+    for dmc in dmc_values:
+        for dc in dc_values:
+            cubes = input_cubes(dmc_val=dmc, dc_val=dc)
+            plugin = BuildUpIndex()
+            plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
            plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+    dmc_data = np.array([[5.0, 10.0, 20.0], [15.0, 25.0, 35.0], [30.0, 45.0, 60.0]])
+    dc_data = np.array([[10.0, 20.0, 40.0], [30.0, 50.0, 70.0], [60.0, 90.0, 120.0]])
+
+    cubes = [
+        make_cube(dmc_data, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_data, "drought_code", "1"),
+    ]
+
+    plugin = BuildUpIndex()
+    plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+    Verifies end-to-end BUI calculation with various input conditions. Tests both
+    equation branches. Edge cases (DMC=0, DC=0, both zero) are covered by dedicated tests.
+
+    Args:
+        dmc_val (float): DMC value to test.
+        dc_val (float): DC value to test.
+        expected_bui (float): Expected BUI output value.
+    """
+    cubes = input_cubes(dmc_val=dmc_val, dc_val=dc_val)
+    result = BuildUpIndex().process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    result = BuildUpIndex().process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+    """
+    dmc_data = np.array([[5.0, 15.0, 30.0], [10.0, 25.0, 45.0], [20.0, 35.0, 60.0]])
+    dc_data = np.array([[10.0, 30.0, 60.0], [20.0, 50.0, 90.0], [40.0, 70.0, 120.0]])
+
+    cubes = [
+        make_cube(dmc_data, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_data, "drought_code", "1"),
+    ]
+
+    result = BuildUpIndex().process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    result = BuildUpIndex().process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+
+    # Calculate expected BUI for each DMC value when DC=0
+    expected_bui = dmc_values - (0.92 + (0.0114 * dmc_values) ** 1.7)
+
+    cubes = [
+        make_cube(dmc_values, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_values, "drought_code", "1"),
+    ]
+
+    result = BuildUpIndex().process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    result = BuildUpIndex().process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+    dc_values = np.array(
+        [[10.0, 50.0, 100.0], [20.0, 75.0, 150.0], [30.0, 90.0, 200.0]]
+    )
+
+    cubes = [
+        make_cube(dmc_values, "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(dc_values, "drought_code", "1"),
+    ]
+
+    result = BuildUpIndex().process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    result = BuildUpIndex().process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+
+    # When DMC=0, BUI should be 0 (no fuel moisture contribution)
+    assert np.allclose(result.data, 0.0, atol=1e-6)
+
+
+def test_process_both_zero() -> None:
+    """Test that when both DMC and DC are zero, BUI equals 0."""
+    cubes = input_cubes(dmc_val=0.0, dc_val=0.0)
+    result = BuildUpIndex().process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    result = BuildUpIndex().process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+def test_input_attribute_mapping() -> None:
+    """Test that input cubes are mapped to correct attribute names.
+
+    Verifies that the base class correctly maps cube standard names to
+    internal attribute names (duff_moisture_code -> input_dmc,
+    drought_code -> input_dc).
+    """
+    cubes = input_cubes(dmc_val=10.0, dc_val=30.0)
+    plugin = BuildUpIndex()
+    plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
    plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py (hunk 12 above)
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
        plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+    elif invalid_input_type == "input_dc_nan":
+        dc_val = np.nan
+    elif invalid_input_type == "input_dc_inf":
+        dc_val = -np.inf
+
+    cubes = input_cubes(dmc_val, dc_val)
+    plugin = BuildUpIndex()
+
+    with pytest.raises(ValueError, match=expected_error):
+        plugin.load_input_cubes(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
        plugin.load_input_cubes(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+    cubes = [
+        make_cube(np.array([[45.9]]), "duff_moisture_code", "1", add_time_coord=True),
+        make_cube(np.array([[123.9]]), "drought_code", "1"),
+    ]
+    plugin = BuildUpIndex()
+
+    # Process should complete without warnings since output stays in valid range
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")  # Turn warnings into errors
+        result = plugin.process(CubeList(cubes))
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]:
```suggestion
        result = plugin.process(cubes)
```

File: improver_tests/fire_weather/test_build_up_index.py [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+import pytest
+from iris.cube import Cube, CubeList
+
+from improver.fire_weather.build_up_index import BuildUpIndex
+from improver_tests.fire_weather import make_cube, make_input_cubes
+
+
+def input_cubes(
+    dmc_val: float = 10.0,
+    dc_val: float = 15.0,
```
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: I think you could use pytest fixtures to simplify these tests:
```suggestion
@pytest.fixture
def input_cubes(
    dmc_val: float,
    dc_val: float,
```

File: improver_tests/fire_weather/test_build_up_index.py (hunk 4 above)
Comment (MoseleyS) [inline improver_tests/fire_weather/test_build_up_index.py]: With the suggested pytest fixture, this would become
```suggestion
def test__calculate(
    dmc_val: float,
    dc_val: float,
    input_cubes,
    expected_bui: float,
) -> None:
    """Test calculation of BUI from DMC and DC.

    Verifies BUI calculation from DMC and DC values.

    Args:
        dmc_val (float): DMC value to test.
        dc_val (float): DC value to test.
        expected_bui (float): Expected BUI value.
    """
    plugin = BuildUpIndex()
    plugin.load_input_cubes(input_cubes)
```

Similar changes would be needed to all the other tests too, and while this might be more pythonic, and also better aligned with other IMPROVER tests, it isn't compulsory.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I've picked up on a few things to think about. The main thing is that I don't think you should include tests of base-class functions in the concrete-class tests.

File: improver/fire_weather/build_up_index.py (hunk 1 above)
Comment (mo-philrelton) [reply improver/fire_weather/build_up_index.py]: I've gone with isclose.

File: improver/fire_weather/build_up_index.py (hunk 2 above)
Comment (mo-philrelton) [reply improver/fire_weather/build_up_index.py:59]: I'm following the equations in the literature. I think this is just a mistake in the FORTRAN implementation.

File: improver/fire_weather/build_up_index.py (hunk 3 above)
Comment (mo-philrelton) [reply improver/fire_weather/build_up_index.py:78]: WIth isclose it could be very close to but below zero, so I'm going to keep this for now.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 25 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: I'm not a fan of using the fixtures. I think it's easier to read as a called function, and means you don't have to pass it as a parameter to each test.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 4 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: Fixed throughout. This required changing the make input cubes method to return tuples rather than lists in order to fix some type checker complaints.

Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: I don't think it's worth it personally. But if anyone else prefers it then we can do so

File: improver_tests/fire_weather/test_build_up_index.py (hunk 5 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: I think you're right, I'll remove it

File: improver_tests/fire_weather/test_build_up_index.py (hunk 6 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:128]: There is: `bui = np.where(both_zero, 0.0, np.where(use_27a, bui_27a, bui_27b))` 

This function converts different sections of the cube to different values based on thresholds. This will vary spatially across the cube if different points meet different thresholds. This should be most of the time, unlike the majority of the tests that just check against single values.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 7 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:198]: As above

File: improver_tests/fire_weather/test_build_up_index.py (hunk 8 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: Ah yes, I've removed both these tests.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 9 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: Good spot, removed.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 10 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: Removed.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 11 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:216]: I think it's best to check that each plugin is set up correctly

File: improver_tests/fire_weather/test_build_up_index.py (hunk 12 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: For this the test is that the values are out of bound for those parameters, in doing so we'd have to import these other modules into the tests for the base class, which I'd like to keep separate. I think it's cleaner to have them as separate tests in this class.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 13 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:285]: See my above answer.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 14 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:313]: That would require importing these classes into the base class test. I think it's cleaner to have it check that the range is correct here. But I have added a check for the warning in the base class test suite.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 6 above)
Comment (MoseleyS) [reply improver_tests/fire_weather/test_build_up_index.py:128]: The IMPROVER repo has acceptance tests for this requirement. My point is that the calculations are independent for each grid point, unlike calculations for the maximum in an area, say.

I don't feel strongly about this, so it can stay as it is, I'm just leaning in the direction of avoiding unit tests for things that aren't the responsibility of the method being tested.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 14 above)
Comment (MoseleyS) [reply improver_tests/fire_weather/test_build_up_index.py:313]: I've looked at this a bit more, and at the base class too, and I think I see what I want changed, but it is a bit fundamental.

I think these plugins would be easier to follow and test if you were strict about class inheritance. What I'm thinking is that the base class has knowledge of the child classes (specifically here, the OUTPUT_RANGES, but also the input units and ranges). There's a bit of blurring here, because I suspect that the input units and ranges overlap between child classes, but for the output ranges, they are all unique and ought to be specified only in the child class that they relate to. If instead, the base class had a `self.output_range = [0, 0]`, which was used in the `self._validate_output_range` method, then all the testing that the range is applied appropriately for a range of ranges, belongs with the base class tests, and the child classes just have to test that they have correctly overridden `self.output_range`.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 25 above)
Comment (MoseleyS) [reply improver_tests/fire_weather/test_build_up_index.py]: I quite like fixtures, but they don't offer much benefit here, so I won't push it.

What I did spot is that the default values for `input_cubes()` are never used, which was what put fixtures in my mind as they do play nicely with `pytest.mark.parametrize`.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 6 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:128]: I agree that we shouldn't be testing for things that we aren't responsible for. But here we're testing that both equations can and are used correctly.

So I think this is a required test for this suite, and is testing the actions of this method rather than the underlying functionality of iris cubes or `numpy.where`. So I'll leave it here for now.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 25 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py]: Thanks, I'll leave it but remove the defaults as suggested.

File: improver_tests/fire_weather/test_build_up_index.py (hunk 14 above)
Comment (mo-philrelton) [reply improver_tests/fire_weather/test_build_up_index.py:313]: Ah this is a good idea. I had been considering doing something similar, but couldn't think of a simple way to do it.

As this would need to be applied to the rest of the FSI plugins, would you mind if I merged this PR as is, and then added this suggestion as a change for all the tests in [this ticket](https://metoffice.atlassian.net/browse/EPPT-3031) I have written for standardising the FSI code?

Comment (MoseleyS) [reply improver_tests/fire_weather/test_build_up_index.py:313]: That's fine.
